### PR TITLE
RL

### PR DESCRIPTION
**Pull request by zikun-li**

### CODE REVIEW

File: .gitignore.orig [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+=======
+experiment/ppo/PPO_logs
+experiment/ppo/PPO_preTrained
+experiment/ppo/pretrain/dataset
+experiment/ppo/pretrain/outputs
+experiment/ppo/pretrain/pretrain_dataset_g9
+experiment/ppo/pretrain/pretrain_dataset_g99
+
+experiment/a2c/*txt
+>>>>>>> Added more files in .gitignore
```
Comment (xumingkuan) [inline .gitignore.orig:86]: Please fix this... Should this file `.gitignore.orig` be removed or replace `.gitignore`?

File: H_CZ_2_2_complete_ECC_set_modified.json [hunk 2]
```diff
@@ -1,11 +0,0 @@
...
-{
-"0_5": [
-[[2,0,0,0,["da715f84238f"],[4.31710838500297678e-01,2.10647530283869677e-01]],[]]
-,[[2,0,0,2,["da715f84238f"],[4.31710838500297567e-01,2.10647530283869650e-01]],[["h", ["Q1"],["Q1"]],["h", ["Q1"],["Q1"]]]]
-,[[2,0,0,2,["da715f84238f"],[4.31710838500297622e-01,2.10647530283869677e-01]],[["h", ["Q0"],["Q0"]],["h", ["Q0"],["Q0"]]]]
-,[[2,0,0,2,["da715f84238f"],[4.31710838500297678e-01,2.10647530283869677e-01]],[["cz", ["Q0", "Q1"],["Q0", "Q1"]],["cz", ["Q0", "Q1"],["Q0", "Q1"]]]]
-,[[2,0,0,2,["da715f84238f"],[4.31710838500297678e-01,2.10647530283869677e-01]],[["cz", ["Q0", "Q1"],["Q0", "Q1"]],["cz", ["Q1", "Q0"],["Q1", "Q0"]]]]
-]
-}
-]
```
Comment (xumingkuan) [inline H_CZ_2_2_complete_ECC_set_modified.json:11]: Have you fixed the cpp code that uses this file?

File: src/quartz/tasograph/tasograph.cpp [hunk 3]
```diff
@@ -1504,25 +1667,13 @@ bool Graph::xfer_appliable(GraphXfer *xfer, Op op) const {
     std::vector<OpX *> output_opxs(num_output, nullptr);
     if (inEdges.find(op_) != inEdges.end()) {
       for (auto &e : inEdges.find(op_)->second) {
+        if (!((size_t)e.dstIdx < num_input)) {
+          std::cout << 1663 << std::endl;
```
Comment (xumingkuan) [inline src/quartz/tasograph/tasograph.cpp:1671]: Is this a debug output? Feel free to merge it as is if it won't output a lot of traces for every run.

File: src/quartz/tasograph/tasograph.cpp [hunk 4]
```diff
@@ -1540,6 +1691,27 @@ bool Graph::xfer_appliable(GraphXfer *xfer, Op op) const {
     if (fail) {
       break;
     }
+    if (outEdges.find(op_) != outEdges.end()) {
+      for (auto &e : outEdges.find(op_)->second) {
+        if (!((size_t)e.srcIdx < num_output)) {
+          std::cout << 1689 << std::endl;
```
Comment (xumingkuan) [inline src/quartz/tasograph/tasograph.cpp:1697]: (same as above)

File: src/quartz/tasograph/tasograph.cpp [hunk 5]
```diff
@@ -1540,6 +1691,27 @@ bool Graph::xfer_appliable(GraphXfer *xfer, Op op) const {
...
+        assert((size_t)e.srcIdx < num_output);
+        output_ops[e.srcIdx] = e.dstOp;
+      }
+    }
+    // Get all output OpX because we don't have output edges in GraphXfer
+    for (auto &opx : xfer->srcOps) {
+      for (auto &input_tensor : opx->inputs) {
+        if (input_tensor.op == opx_) {
+          if (!((size_t)input_tensor.idx < num_output)) {
+            std::cout << 1700 << std::endl;
```
Comment (xumingkuan) [inline src/quartz/tasograph/tasograph.cpp:1708]: (same as above)

File: src/quartz/tasograph/tasograph.cpp [hunk 6]
```diff
@@ -1616,8 +1793,12 @@ bool Graph::xfer_appliable(GraphXfer *xfer, Op op) const {
     opx_op_dq.pop_back();
     xfer->unmatch(opx_op_pair.first, opx_op_pair.second, this);
   }
-  if (!fail)
+  if (!fail) {
+    if (!(mapped_opx.size() == xfer->srcOps.size())) {
+      std::cout << 1790 << std::endl;
```
Comment (xumingkuan) [inline src/quartz/tasograph/tasograph.cpp:1798]: (same as above)

File: src/test/test_apply_xfer.cpp [hunk 7]
```diff
@@ -56,6 +56,16 @@ int main() {
   }
   std::cout << "number of xfers: " << xfers.size() << std::endl;
 
+  size_t cnt = 0;
+  for (size_t i = 0; i < xfers.size(); ++i) {
+    if (xfers[i]->srcOps.size() > xfers[i]->dstOps.size()) {
+      std::cout << ++cnt << ": " << i << std::endl;
+      if (i == 926) {
```
Comment (xumingkuan) [inline src/test/test_apply_xfer.cpp:63]: (same as above)

File: src/test/test_nam_td_disabled.cpp [hunk 8]
```diff
@@ -49,7 +49,8 @@ int main(int argc, char **argv) {
                                        xfer_pair.second, trace);
   auto graph_before_search = graph.toffoli_flip_by_instruction(
       GateType::rz, xfer_pair.first, xfer_pair.second, trace);
-  //   graph_before_search->to_qasm(input_fn + ".toffoli_flip", false, false);
+  graph_before_search->to_qasm(input_fn + ".toffoli_flip", false, false);
+  exit(1);
```
Comment (xumingkuan) [inline src/test/test_nam_td_disabled.cpp:53]: (same as above)

**Review by xumingkuan — COMMENTED** [review]
Thank you for the huge work! I noticed 2 issues below, and I think it would be better to fix them before merging.

If you are debugging at the positions of the other comments I left, feel free to not change them.

File: .gitignore.orig (hunk 1 above)
Comment (zikun-li) [reply .gitignore.orig:86]: I will remove this file

File: H_CZ_2_2_complete_ECC_set_modified.json (hunk 2 above)
Comment (zikun-li) [reply H_CZ_2_2_complete_ECC_set_modified.json:11]: I think this file is still needed in running `running rigetti`.

File: src/quartz/tasograph/tasograph.cpp (hunk 3 above)
Comment (zikun-li) [reply src/quartz/tasograph/tasograph.cpp:1671]: Yep this is a debug output. I'll remove them.

File: src/quartz/tasograph/tasograph.cpp (hunk 4 above)
Comment (zikun-li) [reply src/quartz/tasograph/tasograph.cpp]: Removed

File: src/quartz/tasograph/tasograph.cpp (hunk 5 above)
Comment (zikun-li) [reply src/quartz/tasograph/tasograph.cpp]: Removed

File: src/quartz/tasograph/tasograph.cpp (hunk 6 above)
Comment (zikun-li) [reply src/quartz/tasograph/tasograph.cpp]: Removed

File: src/test/test_apply_xfer.cpp (hunk 7 above)
Comment (zikun-li) [reply src/test/test_apply_xfer.cpp]: Removed

File: src/test/test_nam_td_disabled.cpp (hunk 8 above)
Comment (zikun-li) [reply src/test/test_nam_td_disabled.cpp]: Removed

File: H_CZ_2_2_complete_ECC_set_modified.json (hunk 2 above)
Comment (xumingkuan) [reply H_CZ_2_2_complete_ECC_set_modified.json:11]: So maybe let's restore this file for now.

**Review by xumingkuan — APPROVED** [review]
Feel free to merge this PR at your convenience after restoring the file `H_CZ_2_2_complete_ECC_set_modified.json`.

File: H_CZ_2_2_complete_ECC_set_modified.json (hunk 2 above)
Comment (xumingkuan) [reply H_CZ_2_2_complete_ECC_set_modified.json:11]: I opened issue #28 to track this.

Comment (zikun-li) [reply H_CZ_2_2_complete_ECC_set_modified.json]: Restored.

**Pull request merged**